### PR TITLE
Feature/s h linear regression

### PR DESCRIPTION
**Pull request by sebastianbierb**
- einbinden von Torch.h in MyDataset.h
- Makefile in Ordner src geaendert
- Makefile in Ordner pytorch angepasst fuer die Einbindung von "libtorch"
- in Ordner pytoch TestDataset.cc erstellt und kompilieren funktioniert auch aber bei Ausfuehrung noch
  "Fehlermeldungen -> out of range"
- in Ordner Linearregression ist dann die eigentliche lineare Regression

### CODE REVIEW

**Comment by sebastianbierb** [comment]
Aenderungen:
- State.cc: Konstruktor, Methode calculatescore() , Methode reset()
- Agent.h: Konstruktor, enum Policytype
- Makefile: in pytorch Ordner

**Comment by sebastianbierb** [comment]
Wir haben uns die vorläufige Struktur überlegt in den Dateien:
- LearnedPolicy.h Theta.h RLDataset.h
Würden Sie bitte schauen, ob unsere Ideen sinnvoll sind, bevor wir daran weiter arbeiten.

File: src/ML/SimpleAgent/Agent.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+enum Policytype {rnd, rl};
+/** \brief This Agent class selects edges according to a policy.
+*/
+
+
+class Agent {
+    public:
+
+    
+    //The beginning state for every Episode
```
Comment (MarioStanke) [inline src/ML/SimpleAgent/Agent.h]:
```suggestion
    ///< the initial state where every episode starts
```

File: src/ML/pytorch/LearnedPolicy.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#inlcude "../SimpleAgent/Policy.h"
+#include "../SimpleAgent/state.h"
+
+class LearnedPolicy : public Policy {
+    public:
+        /** This vector changes some process in the act function to change
```
Comment (MarioStanke) [inline src/ML/pytorch/LearnedPolicy.h]: I don't understand.

File: src/ML/pytorch/LearnedPolicy.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+         *  Pi into Pi', vector<float> is a naive idea to change probabilities
+         */
+        vector<float> learnResults;
+
+        /** This function uses theta to calculate v_pi(s, theta) and then
+         *  return learnResults
+         */
+        vector<float> learnFromTheta(vector<float> theta);
```
Comment (MarioStanke) [inline src/ML/pytorch/LearnedPolicy.h]: see my suggestion about Agent in the main review comments

File: src/ML/pytorch/valueMLmodel.cc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+//constructor
+valueMLmodel::valueMLmodel(unsigned int dimstate) {
+    this->dimstate = dimstate;
+    linearNet = std::make_shared<LinearNet>(dimstate);
+}
+
+void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsigned int batch_size, float alpha) {
+    auto data_set = dataSet.map(torch::data::transforms::Stack<>());
+    // create a multi-threaded data loader for the MNIST dataset.
```
Comment (MarioStanke) [inline src/ML/pytorch/valueMLmodel.cc]: wrong comment

File: src/ML/pytorch/valueMLmodel.cc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    auto data_loader = torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(std::move(data_set), batch_size);
+    // instantiate an SGD optimization algorithm to update our LinearNet's parameters.
+    torch::optim::SGD optimizer(linearNet->parameters(), /*lr=*/alpha);
+    for (size_t epoch = 1; epoch <= numberOfEpochs; ++epoch) {
+        torch::Tensor loss;
```
Comment (MarioStanke) [inline src/ML/pytorch/valueMLmodel.cc]: move the initialization outside of the loop

File: src/ML/pytorch/valueMLmodel.cc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+       Returns a boolean vector corresponding to edges in such a way that true means an edge is selectable */
+    vector<bool> index = s->calcSuccessorStates();     
+
+    torch::Tensor succStates = vectorToTensor(s->successorStates);
+
+    // Returns the value prediction for all successor states.       
+    torch::Tensor pred = linearNet->forward(succStates);
+
+    vector<float> tempPrediction = tensorToVector(pred);
+    vector<float> prediction;
```
Comment (MarioStanke) [inline src/ML/pytorch/valueMLmodel.cc]: Initialize with size as the size is known. That is more efficient.

File: src/ML/pytorch/valueMLmodel.cc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    unsigned int counter = 0;
+
+    // To get a vector of the size of the state, we set 0 for every not selectable edge.
+    for (unsigned int i = 0; i < dimstate; i++) {
+        if (index.at(i)) {
```
Comment (MarioStanke) [inline src/ML/pytorch/valueMLmodel.cc]: only a single line required:
```
prediction[i] = index[i]? tempPrediction[counter++] : 0;
```

File: src/ML/pytorch/valueMLmodel.cc [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        for (unsigned int j = 0; j < vec.at(i).size(); j++) {                
+            tens[i][j] = (float)vec.at(i).at(j);
+        }           
+    }
+    return tens;
+}
+
+// Transforms a tensor of values for the successor states into a vector
+vector<float> valueMLmodel::tensorToVector(torch::Tensor& tens) {
+    vector<float*> vec;
```
Comment (MarioStanke) [inline src/ML/pytorch/valueMLmodel.cc]: Please clean up: `vec` is not needed. Also, please initialize vecReturn with size and then do not use push_back.

File: src/ML/pytorch/LearnedPolicy.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+        /* This function maps a state to a probability for each action.
+         * Here it is influenced by learnResults.
+         */
+        void act(state* s);
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/pytorch/LearnedPolicy.h]: I chose runPolicy

File: src/ML/pytorch/LinearNet.h [hunk 10]
```diff
@@ -1,22 +1,27 @@
...
             lin_mod = register_module("fc", torch::nn::Linear(dim_state, 1));
         }
 
         // Implement the Net's algorithm.
-        torch::Tensor forward(torch::Tensor x) {
+        torch::Tensor forward(torch::Tensor& x) {
             return lin_mod->forward(x.reshape({x.size(0), dim_state}));
         }
 
    	    // Use one of many "standard library" modules.
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/pytorch/LinearNet.h]: Didn't find a TAB but i guess that resolves it?

File: src/ML/pytorch/valueMLmodel.cc (hunk 7 above)
Comment (mauricerad) [reply src/ML/pytorch/valueMLmodel.cc]: For consistency within our code and readability we'd like to keep it as is.

File: src/ML/pytorch/LinearNet.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarioStanke) [reply src/ML/pytorch/LinearNet.h]: I am sure there was one. If you don't know how to distinguish tabs from spaces with the editor you use, you can always have a look at the GitHub version of your branch and check the indentation.

[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/pytorch/LinearNet.h]: Alright thanks, we will do that

File: src/ML/SimpleAgent/Agent.h [hunk 11]
```diff
@@ -17,84 +17,48 @@
...
-        episode.actions.push_back(action);
-        episode.numbOfStates = counter + 1;
-        stateAction.first->calculate_score();
-        episode.score = stateAction.first->score;
+    Agent(Graph& g, Policytype pol) : s(g.getEdges()) {
+    switch(pol) {
+        case rnd:
 
-        return episode;
+            policy = new RandomPolicy();
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Agent.h]: We're removing the "new".

File: src/ML/SimpleAgent/Agent.cc [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    stateAction.first->calculate_score();
+	// which is the score for the whole episode.
+    episode.score = stateAction.first->score;
+    
+    return episode;
+}
+
+std::pair <state*, unsigned int> Agent::executePolicy(state* s, Policy* p) {
+	// Create random device.
+    std::random_device rd;
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Agent.cc]: We're haven't found a quick way around this and will be focusing on more impactful efficiency changes first.

File: src/ML/SimpleAgent/Agent.cc [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    std::mt19937 gen(rd());
+	// Get the value estimates for every action of a given state dependant on the policy
+    std::vector <float> probActions = p->runPolicy(s);
+	// Create a distribution dependant on the value estimates
+    std::discrete_distribution<> dis(probActions.begin(), probActions.end());
+	// Randomly choose an action with probabalities of distribution
+    unsigned int edgeSelection = dis(gen);
+    if (s->selectable[edgeSelection] == false) {
+        edgeSelection = -1;
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Agent.cc]: We've followed your suggestion to remove hasEdge() for now. We were aware of this "double check", we were afraid to cause errors we might miss if we remove it though.

[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Agent.cc]: Well we haven't removed hasEdge just the call of hasEdge has been changed to (stateAction.second != -!).

File: src/ML/SimpleAgent/Episode.h [hunk 14]
```diff
@@ -1,15 +1,16 @@
...
 
 class Episode{
     public:
         Episode(){};
-        vector<vector<bool>> states;
-        vector<vector<bool>> actions;
-        unsigned int score;
-        unsigned int numbOfStates;
+        vector<vector<bool>> states;    ///< Vector of SelectedSubsets of states
+        vector<vector<bool>> actions;    ///< Vector of Actions corresponding to states
```
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Episode.h]: We've made an issue for this and will be working on it.

File: src/ML/SimpleAgent/Agent.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (mauricerad) [reply src/ML/SimpleAgent/Agent.h]: Unresolved since there are some issues still

File: src/ML/pytorch/valueMLmodel.cc (hunk 6 above)
Comment (mauricerad) [reply src/ML/pytorch/valueMLmodel.cc]: Will be changed with the efficiency changes for valueMLmodel.

File: src/ML/pytorch/valueMLmodel.cc (hunk 8 above)
Comment (mauricerad) [reply src/ML/pytorch/valueMLmodel.cc]: We're hopefully completely removing both functions and just working with tensors after some talks with Felix.

**Pull request merged**